numpy/lib/npyio.py

```diff
@@ -624,29 +634,15 @@ def _savez(file, args, kwds, compress, allow_pickle=True, pickle_kwargs=None):
     else:
         compression = zipfile.ZIP_STORED
 
-    zipf = zipfile_factory(file, mode="w", compression=compression)
-
-    # Stage arrays in a temporary file on disk, before writing to zip.
-    fd, tmpfile = tempfile.mkstemp(suffix='-numpy.npy')
-    os.close(fd)
-    try:
+    with zipfile_factory(file, mode="w", compression=compression) as zipf:
         for key, val in namedict.items():
-            fname = key + '.npy'
-            fid = open(tmpfile, 'wb')
-            try:
+            with tempfile.NamedTemporaryFile() as fid:
+                fname = key + '.npy'
                 format.write_array(fid, np.asanyarray(val),
                                    allow_pickle=allow_pickle,
                                    pickle_kwargs=pickle_kwargs)
-                fid.close()
-                fid = None
-                zipf.write(tmpfile, arcname=fname)
-            finally:
-                if fid:
-                    fid.close()
-    finally:
-        os.remove(tmpfile)
-
-    zipf.close()
+                fid.flush()
+                zipf.write(fid.name, arcname=fname)
```

Comment:
@charris @juliantaylor that's a problem heavily discuss here:
https://bugs.python.org/issue14243
NamedTemporaryFile is quite unworthy on WinNT because of that issue.

The solution adopted in teh new commit is to have a new context manager, DeleteOnContextExitNamedTemporaryFile
that deletes the file on context exit not on file close.

May be that can be a good idea to externalize the code of that context manager in an other place in the name space, so it can be used elsewhere on the numpy project.